[WIP] Custom status messages

turt2live opened the pull request:
**Requires https://github.com/matrix-org/matrix-js-sdk/pull/805**
Fixes https://github.com/vector-im/riot-web/issues/1528
Comps at https://docs.google.com/document/d/1uLc_KhcWhlQ7r0IyjFLAh2WSV4kJAyIwAN8j6db_VHM/edit#

This is just the UI for status messages. See https://github.com/matrix-org/matrix-js-sdk/pull/805 for details on the technical aspect.

High level features:
* [x] Context menu to change your message
* [x] Status message shown in room list
* [x] Status message shown in member list
* [x] Status message shown in member info

TODO:
* [ ] Improved checkmark icon
* [ ] Bug: Pressing 'clear status' does not clear the input
* [ ] Green border around avatar when a status is set
* [ ] Bug: After setting a status, the clear button doesn't go red/pink

Half-Shot commented:
Is this designed to directly replace `status_msg` being used in the memberlist/being set via a command, because if so then I'm probably gonna need to close #412?

turt2live commented:
No.

From the js-sdk PR:
> Status messages are completely unrelated to the presence status message in that they do not expire and have a different intent. Presence status messages are intended to be read as "Online reading documentation" whereas status messages would be just "Reading documentation" (no associative presence).

Half-Shot commented:
Okay, but the last two features "Status message shown in member list" and "Status message shown in member info" are surely going to conflict with the `status_msg` functionality from a UX pov?

turt2live commented:
Probably, but I don't have mockups for presence status messages and can't make an exception for it :(

Half-Shot commented:
Okay. I'd like for `status_msg` to still work somehow in Riot, but I can see this being a conflicting mess. Maybe there is a way to show both in the UI without causing mass confusion. I'll think about it.

jryans commented on res/css/views/rooms/_RoomTile.scss:
```diff
@@ -35,7 +35,20 @@ limitations under the License.
     display: inline-block;
     width: 180px;
-    height: 24px;
+    //height: 24px;
```
Remove this commented line if not needed.

jryans commented on src/components/views/avatars/MemberStatusMessageAvatar.js:
```diff
@@ -0,0 +1,112 @@
+    }
+}
+
+MemberStatusMessageAvatar.propTypes = {
```
The [style guide](https://github.com/matrix-org/matrix-react-sdk/blob/develop/code_style.md#react) currently recommends using static propTypes inside the class.

jryans commented on src/components/views/avatars/MemberStatusMessageAvatar.js:
```diff
@@ -0,0 +1,112 @@
+        if (ev.getType() !== "im.vector.user_status") return;
+        // TODO: We should be relying on `this.props.member.user._unstable_statusMessage`
+        this.setState({message: ev.getContent()["status"]});
+        this.forceUpdate();
```
What's the reason for this `forceUpdate`? I would expect `setState` to trigger an update itself.

turt2live replied:
Artifact of when I didn't have a `setState`, oops

jryans commented on src/components/views/avatars/MemberStatusMessageAvatar.js:
```diff
@@ -0,0 +1,112 @@
+import * as ContextualMenu from "../../structures/ContextualMenu";
+import StatusMessageContextMenu from "../context_menus/StatusMessageContextMenu";
+
+export default class MemberStatusMessageAvatar extends React.Component {
```
This component appears to depend only its props and state. Could you use [`PureComponent`](https://reactjs.org/docs/react-api.html#reactpurecomponent) here?

jryans commented on src/components/views/context_menus/StatusMessageContextMenu.js:
```diff
@@ -0,0 +1,89 @@
+    }
+}
+
+StatusMessageContextMenu.propTypes = {
```
The [style guide](https://github.com/matrix-org/matrix-react-sdk/blob/develop/code_style.md#react) currently recommends using static propTypes inside the class.

jryans commented on src/components/views/avatars/MemberStatusMessageAvatar.js:
```diff
@@ -0,0 +1,112 @@
+export default class MemberStatusMessageAvatar extends React.Component {
+    constructor(props, context) {
+        super(props, context);
+        this._onRoomStateEvents = this._onRoomStateEvents.bind(this);
```
See the [style guide](https://github.com/matrix-org/matrix-react-sdk/blob/develop/code_style.md#react) for an approach that avoids needing to bind handlers like this.

jryans commented on src/components/views/context_menus/StatusMessageContextMenu.js:
```diff
@@ -0,0 +1,89 @@
+export default class StatusMessageContextMenu extends React.Component {
+    constructor(props, context) {
+        super(props, context);
+        this._onClearClick = this._onClearClick.bind(this);
```
See the [style guide](https://github.com/matrix-org/matrix-react-sdk/blob/develop/code_style.md#react) for an approach that avoids needing to bind handlers like this.

turt2live replied:
I'll still change it, but I really don't like the approach suggested :/

jryans replied:
Ah okay, sorry about that. It's possible we made the style guide change too quickly, as I believe only @ara4n and I discussed it. Code style can certainly be contentious, sometimes with opposing views for certain topics. We may want to use a different process for reviewing code style changes, so that we ensure more people on the team can voice their opinion if they wish to. Any thoughts about this?

turt2live replied:
For now I think it's fine to follow the code style. Just voicing my opinion now so that if/when the code style comes into question again it's not a surprise to people. We probably don't need to iterate on the code style right now anyways, as the part I disagree with doesn't make a difference on maintainability or readability.

jryans commented on src/components/views/avatars/MemberStatusMessageAvatar.js:
```diff
@@ -0,0 +1,112 @@
+    _onRoomStateEvents(ev, state) {
+        if (ev.getStateKey() !== MatrixClientPeg.get().getUserId()) return;
+        if (ev.getType() !== "im.vector.user_status") return;
+        // TODO: We should be relying on `this.props.member.user._unstable_statusMessage`
```
I assume you still need to change this...?

turt2live replied:
Nope, this is a future note. I suppose the js-sdk could emit when it changes, although I'm not sure how worthwhile doing that would be.

Thoughts?

jryans replied:
Ah fair enough, I think I just wasn't following _why_ you couldn't do it now. Maybe extend the comment with ", but there's no specific event to detect when it changes currently." or something similar.

jryans commented:
(I am assuming that Nad will soon try or already has tried your implementation for UX feedback.)

turt2live commented:
> (I am assuming that Nad will soon try or already has tried your implementation for UX feedback.)

Yup :)

jryans commented on src/components/views/context_menus/StatusMessageContextMenu.js:
```diff
@@ -35,7 +35,7 @@ export default class StatusMessageContextMenu extends React.Component {
     }
 
-    _onClearClick = async (e) => {
+    _onClearClick = async(e) => {
```
(This particular linting requirement seems like a bug in our rules, but I am not quite sure where yet...)

turt2live commented:
@jryans hopefully for the last time on this one: please take a look :)

The changes since the last review are:
* https://github.com/matrix-org/matrix-react-sdk/pull/2347/commits/04c9fff6ce9cb4b5ac69adddbac8f709cf6afdb3
* https://github.com/matrix-org/matrix-react-sdk/pull/2347/commits/45f05092ed99a15d28ad3a5b9af57ed2fde1a7a9

jryans commented on src/components/views/avatars/MemberStatusMessageAvatar.js:
```diff
@@ -0,0 +1,120 @@
+        resizeMethod: PropTypes.string,
+    };
+
+    static defaultProps = {
```
Are these default props redundant here? It seems like all 3 of these props flow through to `MemberAvatar` without reading or modifying them, so maybe we don't need to repeat the defaults here as well.

turt2live replied:
They're a little redundant, although I'm not too keen to rely on `MemberAvatar` always having 40/40/crop as its defaults.

jryans replied:
Okay, up to you. I wasn't sure what would be best.

ScottRochford commented:
Has a /slash command been proposed/planned/discarded for setting status messages?
